src/rsp_restspawner/spawner.py

```diff
@@ -3,19 +3,27 @@
 from __future__ import annotations
 
 import asyncio
-from collections.abc import AsyncIterator
+from collections.abc import AsyncIterator, Awaitable, Callable
 from dataclasses import dataclass
 from datetime import timedelta
 from enum import Enum
+from functools import wraps
 from pathlib import Path
-from typing import Any, Optional
+from typing import Any, Optional, TypeVar, cast
 
-from httpx import AsyncClient
+from httpx import AsyncClient, HTTPError
 from httpx_sse import ServerSentEvent, aconnect_sse
 from jupyterhub.spawner import Spawner
 from traitlets import Unicode, default
 
-from .errors import InvalidAuthStateError, MissingFieldError, SpawnFailedError
+from .exceptions import (
+    ControllerWebError,
+    InvalidAuthStateError,
+    MissingFieldError,
+    SpawnFailedError,
+)
+
+F = TypeVar("F", bound=Callable[..., Awaitable[Any]])
```

Comment:
So this is how you do generics in mypy?

Reply:
Yup! It defines a type variable with a class bound and mypy will then complain if you call it with a class that doesn't qualify, but keeps track at the call site of the exact class that you passed in.